docs/source/fx.rst

```diff
@@ -10,7 +10,154 @@ Overview
+are often easier represented through retracing. For example, let’s
+imagine that we wanted to write a decomposition pass that decomposed
+PyTorch functions. For example, it would transform every ``F.relu(x)``
+into ``(x > 0)*x``. One possibility would be to perform the requisite
+graph rewriting to insert the comparison and multiplication after the
+``F.relu``, and then clean up the original ``F.relu``. However, graph
+manipulation can be awkward, and it’s often easier to implicitly
+generate the graph by retracing.
+
+To use this method, we write the graph that we want inserted as regular
+PyTorch code, and pass in implicit proxy objects. These proxy objects
+will capture the operations that are performed on it and append them to
+the graph.
+
+::
+
+    def relu_decomposition(x):
+        return (x > 0)*x
+
+    decomposition_rules = {}
+    decomposition_rules[F.relu] = relu_decomposition
+
+    new_graph = fx.Graph()
+    for node in traced.graph.nodes:
+        if node.op == 'call_function':
+            tracer = GraphAppendingTracer(new_graph)
+            proxy_args = [Proxy(i, tracer=tracer) if isinstance(x, fx.Node)
+                          for x in node.args]
+            new_node = decomposition_rules[node.target](proxy_args)
+            new_node.name = node.name
+        else: # Otherwise, just append the original graph.
+            new_graph.node_copy(node)
+
+In addition to being more natural to write, this also allows you to
+specify your rewrite rules as native Python code. For transformations
+that require a large amount of rewrite rules (such as vmap or grad),
+this can often improve readability and maintainability of the rules.
+
+Example transformations that use this technique include (vmap), (grad),
+and the (decomposition) pass.
+
+FX transforms that don’t return a module
+----------------------------------------
+
+In addition to FX passes (that take in a module and return a module),
+there may be other things you wish to do with the FX graph that don’t
+fit into this paradigm. For example, let’s say that you’d like to obtain
+the shape information of tensors in your graph. In this case, instead of
+looping over the FX graph and modifying it, we can write an interpreter
+on top of the FX graph! As the FX IR is quite simple, it’s easy to
+reimplement an interpreter that also captures your desired attributes.
+
+This example `Shape Propagation
+pass <https://github.com/pytorch/pytorch/blob/master/torch/fx/experimental/shape_prop.py#L7>`__
+reinterprets the FX symbolic graph with example inputs while annotating
+the graph with the shapes.
+
+Reinterpreting the FX graph is generally most useful when you want
+runtime information that FX typically doesn’t capture (due to being a
+symbolic trace). This can be used for capturing shape information for
+downstream passes, but it can also be used to capture other information
+about execution. For example, this (roofline analysis pass in FX)
```

Comment:
I think you wrote that you were formatting the links later, right? Just a reminder comment to go back and do that